File: src/solutions/2024/06/go/main.go
package main

import (
	core "advent/of/go/src/langs/go"
	"bytes"
	"os"
	"strconv"
	"strings"
)

func main() {
	options := core.SolutionOptions{
		HasAlternate: false,
		HasIo:        false,
	}
	core.Run(os.Args, part1, part2, options)
}

func getStart(grid [][]byte) (int, int) {
	for y := 0; y < len(grid); y++ {
		for x := 0; x < len(grid[y]); x++ {
			if grid[y][x] == '^' {
				return x, y
			}
		}
	}
	return 0, 0
}

func part1(input string, isTest bool) string {
	lines := strings.Split(input, "\n")
	grid := make([][]byte, len(lines))
	for i, line := range lines {
		grid[i] = []byte(line)
		grid[i] = append(grid[i], 'x')
		grid[i] = append([]byte("x"), grid[i]...)
	}
	grid = append(grid, bytes.Repeat([]byte("x"), len(grid[0])))
	grid = append([][]byte{bytes.Repeat([]byte("x"), len(grid[0]))}, grid...)

	visited := make([][]bool, len(grid))
	for i := 0; i < len(grid); i++ {
		visited[i] = make([]bool, len(grid[i]))
		for j := 0; j < len(grid[i]); j++ {
			visited[i][j] = false
		}
	}

	directions := [][]int{{0, -1}, {1, 0}, {0, 1}, {-1, 0}}
	dirIndex := 0

	locX, locY := getStart(grid)
	for {
		visited[locY][locX] = true

		nextX := locX + directions[dirIndex][0]
		nextY := locY + directions[dirIndex][1]

		if grid[nextY][nextX] == 'x' {
			break
		}

		if grid[nextY][nextX] == '#' {
			dirIndex = (dirIndex + 1) % 4
		} else {
			locX = nextX
			locY = nextY
		}
	}

	count := int64(0)
	for _, row := range visited {
		for _, val := range row {
			if val {
				count++
			}
		}
	}

	return strconv.FormatInt(int64(count), 10)
}

func part2(input string, isTest bool) string {
	lines := strings.Split(input, "\n")
	grid := make([][]byte, len(lines))
	for i, line := range lines {
		grid[i] = []byte(line)
		grid[i] = append(grid[i], 'x')
		grid[i] = append([]byte("x"), grid[i]...)
	}
	grid = append(grid, bytes.Repeat([]byte("x"), len(grid[0])))
	grid = append([][]byte{bytes.Repeat([]byte("x"), len(grid[0]))}, grid...)

	visited := make([][]bool, len(grid))
	for i := 0; i < len(grid); i++ {
		visited[i] = make([]bool, len(grid[i]))
		for j := 0; j < len(grid[i]); j++ {
			visited[i][j] = false
		}
	}

	directions := [][]int{{0, -1}, {1, 0}, {0, 1}, {-1, 0}}
	dirIndex := 0

	locX, locY := getStart(grid)
	maxCount := int64(0)
	toVisit := [][]int{}
	for {
		if !visited[locY][locX] {
			maxCount++
			toVisit = append(toVisit, []int{locX, locY})
		}
		visited[locY][locX] = true

		nextX := locX + directions[dirIndex][0]
		nextY := locY + directions[dirIndex][1]

		if grid[nextY][nextX] == 'x' {
			break
		}

		if grid[nextY][nextX] == '#' {
			dirIndex = (dirIndex + 1) % 4
		} else {
			locX = nextX
			locY = nextY
		}
	}
	maxCount *= 2

	count := int64(0)
	ogX, ogY := getStart(grid)
	for _, v := range toVisit {
		x, y := v[0], v[1]
		col := grid[y][x]
		if col == 'x' || col == '^' || col == '#' {
			continue
		}

		grid[y][x] = '#'
		locX, locY = ogX, ogY
		dirIndex = 0
		maxIteration := maxCount
		for maxIteration > 0 {
			nextX := locX + directions[dirIndex][0]
			nextY := locY + directions[dirIndex][1]

			if grid[nextY][nextX] == 'x' {
				break
			}

			if grid[nextY][nextX] == '#' {
				dirIndex = (dirIndex + 1) % 4
			} else {
				locX = nextX
				locY = nextY
			}
			maxIteration--
		}
		if maxIteration == 0 {
			count++
		}
		grid[y][x] = '.'
	}

	return strconv.FormatInt(int64(count), 10)
}

File: src/solutions/2024/06/rust/main.rs
extern crate aoc_core;
use std::{env, iter::FromIterator};

fn get_options() -> aoc_core::options::SolutionOptions {
   aoc_core::options::SolutionOptions {
      ..Default::default()
   }
}

fn find_start(grid: &Vec<Vec<char>>) -> (i16, i16) {
   for y in 0..grid.len() {
      for x in 0..grid[0].len() {
         if grid[y][x] == '^' {
            return (x as i16, y as i16);
         }
      }
   }
   (0, 0)
}

fn part_1(input: &str, _is_test: bool) -> String {
   let mut grid: Vec<Vec<char>> = input
      .lines()
      .map(|s| {
         let mut v: Vec<char> = s.chars().into_iter().collect();
         v.insert(0, 'x');
         v.push('x');
         v
      })
      .collect();
   grid.insert(0, vec!['x'; grid[0].len()]);
   grid.push(vec!['x'; grid[0].len()]);
   let mut visited = vec![vec![false; grid[0].len()]; grid.len()];

   let mut loc = find_start(&grid);
   let direction: Vec<(i16, i16)> = vec![(0, -1), (1, 0), (0, 1), (-1, 0)];
   let mut i = 0;
   loop {
      visited[loc.1 as usize][loc.0 as usize] = true;

      let (next_x, next_y) = (loc.0 + direction[i % 4].0, loc.1 + direction[i % 4].1);
      if grid[next_y as usize][next_x as usize] == 'x' {
         break;
      }

      if grid[next_y as usize][next_x as usize] == '#' {
         i = (i + 1) % 4;
      } else {
         loc.0 = next_x;
         loc.1 = next_y;
      }
   }

   visited
      .iter()
      .map(|x| x.iter().filter(|y| **y).count())
      .sum::<usize>()
      .to_string()
}

fn part_2(input: &str, _is_test: bool) -> String {
   let mut grid: Vec<Vec<char>> = input
      .lines()
      .map(|s| {
         let mut v: Vec<char> = s.chars().into_iter().collect();
         v.insert(0, 'x');
         v.push('x');
         v
      })
      .collect();
   grid.insert(0, vec!['x'; grid[0].len()]);
   grid.push(vec!['x'; grid[0].len()]);
   let mut visited = vec![vec![false; grid[0].len()]; grid.len()];

   let mut loc = find_start(&grid);
   let direction: Vec<(i16, i16)> = vec![(0, -1), (1, 0), (0, 1), (-1, 0)];
   let mut i = 0;
   let mut to_visit = vec![];
   loop {
      if !visited[loc.1 as usize][loc.0 as usize] {
         to_visit.push((loc.0 as usize, loc.1 as usize));
      }
      visited[loc.1 as usize][loc.0 as usize] = true;

      let (next_x, next_y) = (loc.0 + direction[i % 4].0, loc.1 + direction[i % 4].1);
      if grid[next_y as usize][next_x as usize] == 'x' {
         break;
      }

      if grid[next_y as usize][next_x as usize] == '#' {
         i = (i + 1) % 4;
      } else {
         loc.0 = next_x;
         loc.1 = next_y;
      }
   }
   let max_visit = to_visit.len() * 2;

   let og_loc = find_start(&grid);
   let mut count = 0;
   for v in to_visit {
      let (x, y) = v;
      if !visited[y][x] || grid[y][x] == '#' || grid[y][x] == '^' {
         continue;
      }
      grid[y][x] = '#';

      loc.0 = og_loc.0;
      loc.1 = og_loc.1;
      i = 0;
      let mut max_iter = max_visit;
      while max_iter > 0 {
         let (next_x, next_y) = (loc.0 + direction[i % 4].0, loc.1 + direction[i % 4].1);
         if grid[next_y as usize][next_x as usize] == 'x' {
            break;
         }

         if grid[next_y as usize][next_x as usize] == '#' {
            i = (i + 1) % 4;
         } else {
            loc.0 = next_x;
            loc.1 = next_y;
         }

         max_iter -= 1;
      }
      if max_iter == 0 {
         count += 1;
      }
      grid[y][x] = '.';
   }

   count.to_string()
}

fn main() {
   let args = Vec::from_iter(env::args());
   aoc_core::run(
      args.iter().map(|x| x.as_ref()).collect::<Vec<_>>(),
      part_1,
      part_2,
      get_options(),
   );
}
